Fix scrolling behaviour on page navigation

amanda-anderson opened the pull request:
# Description
### Issue 1
After some investigation I realized that the weird scrolling behaviour when you load a new page (and it sometimes sets you halfway down the page) was because the page had been visited previously. Also since React only knows that there are new components on the screen (but not that we should be scrolling to the top when they are displayed) ,this happens on all pages unless we add code to scroll to the top manually. 

I originally ended up looking into React Router here, since I figured that on all new routes we want to start from the top by default. In some cases we _do_ want to start the page scrolled down, for example in idea lists when we click on an idea and then want to go back to where we were in the list.

After lots of reading online, I think the best solution is for each clHistory.push, if we want the page to be scrolled up we should specify that directly using a new function scrollToTop(). This is the best way to control in what exact contexts the scroll up happens, and also still allows the browser to make the best decision for other cases (E.g. going back to previous URLs etc.). Browsers have a lot of built in functionality for this, so I think it's best not to override this unless we're sure :slightly_smiling_face: 

### Issue 2
When you visit a project page with the idea list/map view toggle, the user would get automatically scrolled to this toggle. It looks like it was related to an accessibility fix we made, however it ends up focusing the user on this toggle on render. I originally tried to replace the current ref implementation with useState, however that didn't work. I ended up adding a simple useState to keep track of when the view was changed instead, and only trigger the onFocus when the view has been changed for the first time.

### Issue 3
Another scrolling issue came from the Read More button on idea cards. I think the readMore "linkTo" implementation doesn't play very nicely with the browser back button in the idea page :thinking: If you clicked "read more" on the card and then clicked the browser back button, it would randomly scroll the user down the page rather than go back. To fix this, I implemented the readMore like our goBack button instead, where it accepts either an onClick function or a linkTo. I hope this is alright? :thinking: It does seem to fix the problem.

# Changelog
## Fixed
- Fixed a number of resident-facing scrolling issues when navigating platform:
    - When visiting pages which had been opened previously, the page would load in the same scroll position of the current page (I.e. New page could load halfway down the page which was very confusing).
    -  Project pages which had an idea map view automatically scrolled to the map/list view toggle when they loaded.
    - Clicking back button on idea page sometimes scrolled down to comment section rather than going back to previous page.

amanda-anderson committed:
Fix bad scrolling behaviour on page navigation

amanda-anderson committed:
Fix scroll issue on goBack from idea page when read more was clicked on idea card

amanda-anderson committed:
Fix issue with scrolling to idea list

IvaKop reviewed (approved):
LGTM! As long as you tested it well, I think it's good to go :)

amanda-anderson committed:
Test timeout

amanda-anderson committed:
Test timeout

amanda-anderson commented:
@IvaKop Just a quick update on this one - there was indeed one case I missed while testing. Just looking at another potential solution now :) Depending on how different it becomes I might re-assign you to review once it's implemented.

amanda-anderson committed:
Change scrollToTop implementation

amanda-anderson committed:
Merge with master

amanda-anderson committed:
Rename function

amanda-anderson committed:
Use timeout

amanda-anderson committed:
Fix variable name

amanda-anderson committed:
Remove import

amanda-anderson committed:
Handle scroll on project card click

amanda-anderson committed:
Use timeout only for mobile

amanda-anderson commented:
@IvaKop I've done quite a bit of testing + reading, and I've changed the approach for Issue # 1 a bit. We can't make the assumption that a URL change should result in scrolling to the top. For cases like going back, all browsers have built in handling for this which works well to maintain previous scroll positions, so I don't want to disrupt that either. The recommended solution I've seen online is to add a scrollToTop function which we can add in for the exact page transitions we want it for. 

So I've gone in and added this for a number of page transitions in our front office. The behaviour is now much more predictable than the automatic scrolling I added in earlier. I also noticed during testing on my own device + reading online that for some mobile browsers we need to use the scroll to top in a timeout function for it to function properly.

amanda-anderson commented on front/app/utils/scroll.ts:
```diff
@@ -25,3 +27,17 @@ export function scrollToElement({
+  if (isMobileOrSmaller) {
+    setTimeout(() => {
+      window.scrollTo(0, 0);
+    }, 100);
```
I tested on my physical device (on an epic deployment), and indeed this timeout was necessary. Since this is a function I can't use our `useBreakpoint` hook, but I hope using the `viewportWidths` directly is fine?

IvaKop replied:
I think it's not so bad, as long as we use the constants from the component library 👍

IvaKop commented on front/app/api/users/useDeleteSelf.ts:
```diff
@@ -35,6 +36,7 @@ const useDeleteSelf = () => {
         invalidateQueryCache();
       }
       clHistory.push('/');
+      scrollToTop();
```
I wonder if, instead of calling this util separately, we shouldn't add it as an option to the `clHistory.push` and `clHistory.replace` methods. It would make more sense to have this functionality there as an option, I think.

IvaKop commented on front/app/components/Avatar/index.tsx:
```diff
@@ -142,7 +143,12 @@ const Avatar = memo(
     if (isLinkToProfile && hasValidProfileLink) {
       return (
-        <Link to={profileLink}>
+        <Link
```
It is not a good idea to have a path and onClick on a link component at the same time, it feels very confusing. What if, instead, we add a `scrollToTop` prop to the link component that handles this functionality in the component itself.

IvaKop commented on front/app/components/IdeaCard/Footer/ReadMoreButton.tsx:
```diff
@@ -6,10 +6,11 @@ import messages from '../messages';
 interface Props {
-  slug: string;
+  slug?: string;
+  onClick?: () => void;
```
Here also, I would not add an additional `onClick` handler but manage the scroll to top behaviour in the component.

IvaKop commented on front/app/components/InitiativeCard/index.tsx:
```diff
@@ -93,6 +94,9 @@ const InitiativeCard = ({
     <Card
       className={cardClassNames}
       to={`/initiatives/${initiative.data.attributes.slug}?go_back=true`}
+      onClick={() => {
```
Same comment as above.

IvaKop commented on front/app/components/InitiativesMap/InitiativePreview.tsx:
```diff
@@ -170,6 +171,9 @@ const InitiativePreview = ({
       <ViewInitiativeButton
         fullWidth={true}
         linkTo={`/initiatives/${initiative.data.attributes.slug}?go_back=true`}
+        onClick={() => {
```
Same comment as above.

IvaKop commented on front/app/components/ProjectCard/index.tsx:
```diff
@@ -445,14 +448,17 @@ const ProjectCard = memo<InputProps>(
 
     const handleProjectCardOnClick = (projectId: string) => () => {
       trackEventByName(tracks.clickOnProjectCard, { extra: { projectId } });
+      scrollToTop();
```
In cases like these, let's handle the scroll behaviour where the actual navigation happens, in this case in the `styled(Link)` container.

IvaKop commented on front/app/utils/scroll.ts:
```diff
@@ -25,3 +27,17 @@ export function scrollToElement({
+
+  // On some mobile browsers, we need a timeout for this to work correctly
+  if (isMobileOrSmaller) {
+    setTimeout(() => {
```
Let's also clear this timeout.

IvaKop reviewed (changes requested):
I think the overall approach is good but I would suggest some changes to ensure it's more clear where and how we handle scroll behaviour.
- Adding a scroll to top option to `clHistory.push` and `clHistory.replace` methods, instead of calling a separate method for this
- Adding `scrollToTop` functionality to the link component, instead of creating `onClick` handlers for it
- Managing other components with navigation functionality in a similar way to the `Link` component, instead of using `onClick` handlers.

amanda-anderson committed:
Changes from code review

amanda-anderson committed:
Changes from code review

amanda-anderson committed:
Fix from code review

amanda-anderson committed:
Use timeout only for links

amanda-anderson committed:
Fix onClick

amanda-anderson committed:
Remove unnecessary code

amanda-anderson committed:
Remove unnecessary prop

amanda-anderson committed:
Remove unnecessary scrollToTop

amanda-anderson committed:
Add scrollToTop back into card

IvaKop commented on front/app/utils/cl-router/history.ts:
```diff
@@ -1,29 +1,37 @@
   method: 'push' | 'replace',
-  location: Partial<Location> | string
+  location: Partial<Location> | string,
+  scrollToTop?: boolean
```
Let's make this an object where one of the keys is `scrollToTop`. This will make it easier if we want to add new settings in the future + would make the use of this setting much more explicit. So instead of `clHistory.push('/', true)` you would have `clHistory.push('/', { scrollToTop: true })`

amanda-anderson replied:
@IvaKop Ok sounds good - yeah much more clear that way :+1: Do you mean something like this then? Or do we have a different preferred format? :thinking: 

```
type ClHistoryProps = {
  scrollToTop?: boolean;
};

// overrides push and replace methods so they update the location with the current locale from the locale stream
function historyMethod(
  method: 'push' | 'replace',
  location: Partial<Location> | string,
  props?: ClHistoryProps
): void {
  // 'gets' current locale
  localeStream()
    .observable.pipe(first())
    .subscribe((locale) => {
      // calls the vanilla react-router method with updated location
      history[method](updateLocationDescriptor(location, locale));

      if (props?.scrollToTop) {
        scrollTop();
      }
    });
}

export default {
  ...history,
  push: (location: Partial<Location> | string, props?: ClHistoryProps): void =>
    historyMethod('push', location, { scrollToTop: props?.scrollToTop }),
  replace: (
    location: Partial<Location> | string,
    props?: ClHistoryProps
  ): void =>
    historyMethod('replace', location, { scrollToTop: props?.scrollToTop }),
  goBack: () => history.back(),
};
```

IvaKop replied:
Yes, exactly! Although I would probably name this `options` instead of props as I associate props more with a React component.

amanda-anderson replied:
Ahh true - ok I'll change the name :) Thanks Iva!

IvaKop commented on front/app/containers/MainHeader/DesktopNavbar/AdminPublicationsNavbarItem.tsx:
```diff
@@ -209,6 +209,7 @@ const AdminPublicationsNavbarItem = ({
                   {item.relationships.publication.data.type === 'project' && (
                     <ProjectsListItem
                       to={`${linkTo}/${item.attributes.publication_slug}`}
+                      scrollToTop={true}
```
You can use the short version (without `={true}`) but this is fine, too.

IvaKop reviewed (changes requested):
Everything looks super nice now! 

One last small change request - we need to change the way we pass `scrollToTop` in the `clHistory` methods just to make it more explicit and easy for the dev team to understand how to use this. I've described the change in detail in a comment below.

amanda-anderson committed:
Fixes from code review

amanda-anderson committed:
Use new options prop for scrollToTop clHistory

amanda-anderson commented:
The 2 failing e2e tests were already failing on master. All others are passing.
